### ISSUE TITLE
Cilium CNI IPAM crd mode: No more IPs available

### ISSUE THREAD

**Issue by houminz** [cilium/cilium#17691]
## Bug report

**General Information**

- Cilium version (run `cilium version`) v1.10.4
- Kernel version (run `uname -a`) linux 5.4.32
- Orchestration system version in use (e.g. `kubectl version`, ...)
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.4-tke.11", GitCommit:"2093e9ec8db4fd321acac1e8b78ba28792ff6395", GitTreeState:"clean", BuildDate:"2021-06-04T14:17:35Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.4-tke.11", GitCommit:"2093e9ec8db4fd321acac1e8b78ba28792ff6395", GitTreeState:"clean", BuildDate:"2021-06-04T14:12:37Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
```
- Link to relevant artifacts (policies, deployments scripts, ...)
```
quick install script
```
- Generate and upload a system zip:
```
kubectl exec -it cilium-8lzm6 -nkube-system -- cilium debuginfo > sysdump.txt
```
[sysdump.txt](https://github.com/cilium/cilium/files/7409563/sysdump.txt)

**How to reproduce the issue**

1. patch CiliumNode object  `spec.ipam.pool`  so that there are more ip to use
```
spec:
  addresses:
  - ip: 10.19.1.13
    type: InternalIP
  - ip: 10.200.0.13
    type: CiliumInternalIP
  - ip: 10.200.0.29
    type: CiliumInternalIP
  azure: {}
  encryption: {}
  eni: {}
  health:
    ipv4: 10.200.0.30
  ipam:
    podCIDRs:
    - 10.200.0.8/29
    - 10.200.0.24/29
    - 10.200.0.48/29
    pool:
      10.200.0.9: {}
      10.200.0.10: {}
      10.200.0.11: {}
      10.200.0.12: {}
      10.200.0.13: {}
      10.200.0.14: {}
      10.200.0.25: {}
      10.200.0.26: {}
      10.200.0.27: {}
      10.200.0.28: {}
      10.200.0.29: {}
      10.200.0.30: {}
      10.200.0.49: {}
      10.200.0.50: {}
      10.200.0.51: {}
      10.200.0.52: {}
      10.200.0.53: {}
      10.200.0.54: {}
status:
  alibaba-cloud: {}
  azure: {}
  eni: {}
  ipam:
    operator-status: {}
    used:
      10.200.0.9:
        owner: default/busybox-65df9f457b-drsx4
      10.200.0.10:
        owner: default/busybox-65df9f457b-bp785
      10.200.0.11:
        owner: default/busybox-ggmpf [restored]
      10.200.0.12:
        owner: default/busybox-65df9f457b-vgd8q
      10.200.0.13:
        owner: default/busybox-65df9f457b-26794
      10.200.0.14:
        owner: default/busybox-65df9f457b-jrc9c
      10.200.0.25:
        owner: default/busybox-65df9f457b-5zlxl
      10.200.0.26:
        owner: default/busybox-65df9f457b-lmbpt
      10.200.0.27:
        owner: default/busybox-65df9f457b-47bhn
      10.200.0.28:
        owner: default/busybox-65df9f457b-gpjs6
      10.200.0.29:
        owner: router
      10.200.0.30:
        owner: health
      10.200.0.49:
        owner: default/busybox-65df9f457b-m7x5z
      10.200.0.50:
        owner: default/busybox-65df9f457b-xc65k
```
2. scale deployment to larger size
```
$ kubectl get po -owide | grep busybox-65df9f457b | grep node-10.19.1.13
busybox-65df9f457b-26794   1/1     Running             0          115m    10.200.0.13    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-47bhn   1/1     Running             0          115m    10.200.0.27    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-4k486   1/1     Running             0          113m    10.200.0.52    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-5v2md   1/1     Running             0          113m    10.200.0.51    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-5zlxl   1/1     Running             0          115m    10.200.0.25    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-7gjwx   0/1     ContainerCreating   0          113m    <none>         node-10.19.1.13   <none>           1/1
busybox-65df9f457b-bdfx2   0/1     ContainerCreating   0          113m    <none>         node-10.19.1.13   <none>           1/1
busybox-65df9f457b-bp785   1/1     Running             0          115m    10.200.0.10    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-drsx4   1/1     Running             0          115m    10.200.0.9     node-10.19.1.13   <none>           1/1
busybox-65df9f457b-gpjs6   1/1     Running             0          115m    10.200.0.28    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-jrc9c   1/1     Running             0          115m    10.200.0.14    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-l8ntv   1/1     Running             0          113m    10.200.0.54    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-lmbpt   1/1     Running             0          115m    10.200.0.26    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-m7x5z   1/1     Running             0          115m    10.200.0.49    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-vgd8q   1/1     Running             0          115m    10.200.0.12    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-xc65k   1/1     Running             0          115m    10.200.0.50    node-10.19.1.13   <none>           1/1
busybox-65df9f457b-xtdth   1/1     Running             0          113m    10.200.0.53    node-10.19.1.13   <none>           1/1
```
3. we can see from the CiliumNode that there are still some IP to be allocated, however, the pod busybox-65df9f457b-7gjwx could not up

```
$ kubectl describe po busybox-65df9f457b-7gjwx

Events:
  Type     Reason                  Age                     From                      Message
  ----     ------                  ----                    ----                      -------
  Warning  FailedCreatePodSandBox  8m2s (x5806 over 112m)  kubelet, node-10.19.1.13  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "e66d80d2c5a719db27d21100eb8f7bab63d8c832c296e84924c08013c2040c01" network for pod "busybox-65df9f457b-7gjwx": networkPlugin cni failed to set up pod "busybox-65df9f457b-7gjwx_default" network: cmdAdd: Err in setup plugins: delegateAdd: error in invoke Conf add - "cilium-cni": error in getting result from AddNetwork: unable to allocate IP via local cilium agent: [POST /ipam][502] postIpamFailure  No more IPs available
```

**Comment by alex-berger** [comment]
We also  observe this error message from time-to-time, but without much negative impact (in the end the Pods are running).

<img width="845" alt="image" src="https://user-images.githubusercontent.com/1087375/145799108-c08198e4-7426-473d-bf6f-d8803a665e0e.png">

**Comment by christarazi** [comment]
```
level=warning msg="Unable to update CiliumNode custom resource" error="Operation cannot be fulfilled on ciliumnodes.cilium.io \"node-10.19.1.13\": the object has been modified; please apply your changes to the latest version and try again" subsys=ipam
```

The fact that we see these msgs repeated is an indication that the `nodeStore` does not have the most up-to-date resource. The fix would to understand why that's the case. Otherwise the actual IPAM logic should just work.

**Comment by christarazi** [comment]
@gandro Given that you've been working on the [cluster-pool-v2 IPAM](https://github.com/cilium/cilium/pull/18887) mode, cc'ing you for relevance. No need for you to investigate this issue.

```
level=warning msg="Unable to update CiliumNode custom resource" error="Operation cannot be fulfilled on ciliumnodes.cilium.io \"node-10.19.1.13\": the object has been modified; please apply your changes to the latest version and try again" subsys=ipam
```

When the CiliumNode IPAM pool is full and pods are deleted (IPAM deallocate), it seems that Cilium could stuck in a place where the local node resource is out-of-date (some kind of race maybe?). In case you see this in your testing of cluster-pool-v2, just FYI.

**Comment by gandro** [comment]
@christarazi Yes, I encountered that bug too, this is why we have commit https://github.com/cilium/cilium/pull/18887/commits/777b1f0256bd9350c09fd9f4febd825dce1decfc in clusterpoolv2.

However, I already fixed that same in IPAM:CRD a while ago: https://github.com/cilium/cilium/pull/17856 - however, this is only available in v1.11 at the moment.

**Comment by christarazi** [comment]
@SimpCosm Can you try v1.11.x?
